druid-shell/src/platform/x11/dbus.rs

```diff
@@ -0,0 +1,264 @@
+//! This module contains functions for opening file dialogs using DBus.
+
+use std::collections::HashMap;
+use std::os::unix::ffi::OsStringExt;
+
+use tracing::warn;
+use zbus::{dbus_proxy, fdo, Connection};
+use zvariant::{OwnedObjectPath, OwnedValue};
+
+use super::window::{IdleHandle, IdleKind};
+use crate::{FileDialogOptions, FileDialogToken, FileInfo};
+
+// I made this struct to store, e.g., connection state and pending responses. But I'm getting
+// deadlocks when trying to use the connection, so for now just make a new connection each time.
+#[derive(Clone, Default)]
+pub struct DbusHandle {}
+
+fn response_thread(
+    conn: Connection,
+    idle: IdleHandle,
+    path: OwnedObjectPath,
+    tok: FileDialogToken,
+    open: bool,
+) -> fdo::Result<()> {
+    let path = path.into_inner();
+    loop {
+        let msg = conn.receive_message()?;
+        let msg_header = msg.header()?;
+        if msg_header.message_type()? == zbus::MessageType::Signal
+            && msg_header.member()? == Some("Response")
+        {
+            if let Some(reply_path) = msg.header()?.path()? {
+                if &path == reply_path {
+                    let response = msg.body::<ResponseResult<FileDialogResponse>>()?;
+                    let file_info = if let ResponseType::Success = response.0 {
+                        // FIXME: what are we supposed to do if there are multiple files?
+                        response.1.uris.first().and_then(|s| {
+                            // FIXME: we get a URI, but we want a path. There must be a less hacky
+                            // way to do this...
+                            if s.starts_with("file://") {
```

Comment:
[strip_prefix](https://doc.rust-lang.org/std/primitive.str.html#method.strip_prefix)?

That's what I use in my editor for handling URIs from LSP servers but my understanding is that there are other complications that can arise